### ISSUE TITLE
Controlled recursion for rules to avoid stack overflow

### ISSUE THREAD

**Issue by refset** [xtdb/crux#1930]
I expect that there are number of ways this requirement could be interpreted and implemented.

One possible userspace workaround would be to thread a reverse counter var through the rule heads, `dec` in each rule body and stop when you hit `0`, similar to: https://gist.github.com/refset/57a910e2b746332ec7a0a31886f57cfb#file-better-tree-depth-clj-L14-L15

**Comment by djtango** [comment]
@refset yes that's what I ended up going with - a counter for the recur depth :+1:
